src/buffer/out/textBuffer.cpp

```diff
@@ -1146,6 +1150,41 @@ void TextBuffer::Reset() noexcept
     _initialAttributes = _currentAttributes;
 }
 
+void TextBuffer::ClearScrollback(const til::CoordType start, const til::CoordType height)
+{
+    if (start <= 0)
+    {
+        return;
+    }
+
+    if (height <= 0)
+    {
+        _decommit();
+        return;
+    }
+
+    // Our goal is to move the viewport to the absolute start of the underlying memory buffer so that we can
+    // MEM_DECOMMIT the remaining memory. _firstRow is used to make the TextBuffer behave like a circular buffer.
+    // The start parameter is relative to the _firstRow. The trick to get the content to the absolute start
+    // is to simply add _firstRow ourselves and then reset it to 0. This causes ScrollRows() to write into
+    // the absolute start while reading from relative coordinates. This works because GetRowByOffset()
+    // operates modulo the buffer height and so the possibly-too-large startAbsolute won't be an issue.
+    const auto startAbsolute = _firstRow + start;
+    _firstRow = 0;
+    ScrollRows(startAbsolute, height, -startAbsolute);
+
+    const auto end = _estimateOffsetOfLastCommittedRow();
+    for (auto y = height; y < end; ++y)
+    {
+        GetMutableRowByOffset(y).Reset(_initialAttributes);
+    }
+
+    // Clear all the marks below the new viewport position.
+    ClearMarksInRange(til::point{ 0, height }, til::point{ _width, _height });
+    // Then scroll all the remaining marks up. This will trim ones that are now "outside" the buffer
+    ScrollMarks(-start);

```

Comment:
Should this not be `-startAbsolute` now that we're moving to absolute 0?

Also, the order of these two operations seems wrong to me. I know that's how it's currently implemented, but surely that can't have been correct? If you're clearing from the `height` offset, that would assume the positions have already been moved to the top of the buffer.

Reply:
I'm not thoroughly familiar with the marks implementation just yet, but I believe it's effectively "sideband" data that is entirely detached from the other `TextBuffer` internals. The marks coordinates are in absolute buffer coordinates (like console API coordinates) and so they don't need to be adjusted. You can see this in `Terminal::NotifyBufferRotation` as it calls `ScrollMarks` to adjust all marks coordinates on every newline to keep them in absolute offsets.

And yeah I was wondering about the order of operations as well, but I didn't think about it too deeply. Now that I look at it again, I realize that they're indeed out of order, as you've said.

Reply:
Actually I just realize that I sort of mixed absolute and relative. Their meaning swaps depending on whether we're talking about console API coordinates or the underlying memory offsets... I hope my explanation was intelligible (?) despite this. 😅

Reply:
OK, I wasn't familiar with the marks implementation either, but what you said make sense.